### PR TITLE
Macro Preparation

### PR DESCRIPTION
**Pull request by iamrecursion**
### Pull Request Description
This PR includes the preparation for the creation of the macro resolver.

### Important Notes
This does not actually implement the token stream chunking.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/main/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/main/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/main/docs/style-guide/rust.md) style guides.
- [x] All documentation and configuration conforms to the [markdown](https://github.com/enso-org/enso/blob/main/docs/style-guide/markdown.md) and [YAML](https://github.com/enso-org/enso/blob/main/docs/style-guide/yaml.md) style guides.
- [x] All code has been tested where possible.

### CODE REVIEW

File: lib/rust/parser/benches/bench_tree_query.rs [hunk 1]
```diff
@@ -0,0 +1,99 @@
+//! This file contains benchmarks of the query performance for the tree structure used in the macro
+//! resolver.
+
+use parser::data::tree::*;
+use itertools::*;
+
+use criterion::{criterion_group, criterion_main, Criterion, black_box};
```
Comment (wdanilo) [inline lib/rust/parser/benches/bench_tree_query.rs]: were not using such imports. Please use single line imports instead

File: lib/rust/parser/benches/bench_tree_query.rs [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    }));
+}
+
+fn deep_tree(c:&mut Criterion) {
+    let tree  = gen_tree(5,10);
+    let query = &[1,2,4,1,3];
+    c.bench_function("Deep Tree",|b| b.iter(|| {
+        tree.get_value(black_box(query))
+    }));
+}
```
Comment (wdanilo) [inline lib/rust/parser/benches/bench_tree_query.rs]: I need soe docs here describing what structure are these generating

File: lib/rust/parser/src/data/tree.rs [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+pub type Branches<K,V> = HashMap<K,Tree<K,V>>;
+
+/// A tree with a variable number of branches per node.
+#[derive(Derivative)]
+#[derivative(Clone)]
+#[derivative(Debug(bound="K:Debug+Eq+Hash+PartialEq, V:Debug+Eq+PartialEq"))]
+#[derivative(Default(bound="K:Eq+Hash"))]
+#[derivative(PartialEq(bound="K:Eq+Hash, V:PartialEq"))]
+#[derivative(Eq(bound="K:Eq+Hash, V:Eq"))]
+pub struct Tree<K,V> {
```
Comment (wdanilo) [inline lib/rust/parser/src/data/tree.rs]: I would name it HashTree or HashMapTree

File: lib/rust/parser/src/data/tree.rs [hunk 4]
```diff
@@ -0,0 +1,214 @@
+//! A map-based tree for use in the macro resolver.
```
Comment (wdanilo) [inline lib/rust/parser/src/data/tree.rs]: Can you move this struct out of parser to our common libs please? Its super useful

File: lib/rust/parser/src/data/tree.rs [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+        self.branches.is_empty()
+    }
+
+    /// Check if `self` is a non-leaf node in the tree.
+    pub fn is_non_leaf(&self) -> bool {
+        !self.is_leaf()
+    }
+}
+
+impl<K,V> Tree<K,V> where K:Clone+Eq+Hash+PartialEq {
```
Comment (wdanilo) [inline lib/rust/parser/src/data/tree.rs]: Create alias for the types, like `trait Bounds = Clone+Eq+Hash+PartialEq;` which can be accessed fro the outside as `tree::Bounds` - that will make external code much nicer sometimes

File: lib/rust/parser/src/data/tree.rs [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+    /// Create an empty tree.
+    pub fn empty() -> Self {
+        default()
+    }
```
Comment (wdanilo) [inline lib/rust/parser/src/data/tree.rs]: I dont like that for some things, like `is_leaf` you dont require bounds (and thatas cool, as you require minimal set of them), and here you require all of them. Each function should require only those bounds that are really needed there

File: lib/rust/parser/src/data/tree.rs [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+
+    /// Create a singleton tree.
+    pub fn singleton(value:V) -> Self {
+        let value    = Some(value);
+        let branches = default();
+        Self{value,branches}
+    }
+
+    /// Insert the provided `value` into the tree at the provided `path`.
+    pub fn insert(&mut self, path:&[K], value:V) {
```
Comment (wdanilo) [inline lib/rust/parser/src/data/tree.rs]: I would do `I : IntoIter<Item : Into<K>>` and `path : I` instead

File: lib/rust/parser/src/data/tree.rs [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+    }
+
+    /// Map the provided `f` over `self`, mutating the tree.
+    ///
+    /// This may change the value type stored in the tree.
+    ///
+    /// ## NOTE
+    /// This function is only suitable for use on trees with small depths as it is implemented in a
+    /// recursive fashion.
+    pub fn map<S,F:Copy+Fn(V) -> S>(self, f:F) -> Tree<K,S> {
```
Comment (wdanilo) [inline lib/rust/parser/src/data/tree.rs]: With bigger bounds, like here, lets use `where` clause on second line instead

File: lib/rust/parser/src/data/tree.rs [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+        let value         = self.value.map(f);
+        let branches_iter = self.branches.into_iter().map(|(k,v)| (k,v.map(f)));
+        let branches      = branches_iter.collect::<Branches<K,S>>();
```
Comment (wdanilo) [inline lib/rust/parser/src/data/tree.rs]: You dont need to provide type, do you?

File: lib/rust/parser/src/data/tree.rs [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+        Tree{value,branches}
```
Comment (wdanilo) [inline lib/rust/parser/src/data/tree.rs]: USe Self! Less refactoring in the future :)

File: lib/rust/parser/src/data/tree.rs [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+    }
+
+    /// Map the provided `f` over `self`, mutating the tree in place.
```
Comment (wdanilo) [inline lib/rust/parser/src/data/tree.rs]: Use correct links in docs please. We were using wring style. If Im reading docs correctly, it should be ```[`f`]``` instead. See https://rust-lang.github.io/rfcs/1946-intra-rustdoc-links.html

File: lib/rust/parser/src/macros.rs [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+pub mod literal;
+pub mod registry;
+
+use crate::prelude::*;
+use crate::prelude::logger::*;
+
+use crate::macros::definition::Definition;
+use crate::macros::registry::Registry;
+
+
```
Comment (wdanilo) [inline lib/rust/parser/src/macros.rs:13]: 3 nlns here

File: lib/rust/parser/src/macros.rs [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+/// The Enso macro resolver.
+#[derive(Clone,Debug,PartialEq)]
+pub struct Resolver<Logger> {
+    /// The macro registry.
+    registry : Registry,
+    /// The logger for the macro resolver.
+    logger : Logger
+}
+
+impl<Logger:AnyLogger<Owned=Logger>> Resolver<Logger> {
```
Comment (wdanilo) [inline lib/rust/parser/src/macros.rs]: Move that to where clause in next line. We try to keep the things on the left minimal and easy to read

File: lib/rust/parser/src/macros.rs [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+// === Resolver ===
+// ================
+
+/// The Enso macro resolver.
+#[derive(Clone,Debug,PartialEq)]
+pub struct Resolver<Logger> {
+    /// The macro registry.
+    registry : Registry,
+    /// The logger for the macro resolver.
+    logger : Logger
```
Comment (wdanilo) [inline lib/rust/parser/src/macros.rs]: remove these docs. Doc telling that "logger" is a logger" and "macro::Registry" is a macro registry are useless. Add allow no docs and it iwll be cleaner

File: lib/rust/parser/src/macros.rs [hunk 15]
```diff
@@ -0,0 +1,40 @@
+//! The macro system for the Enso parser.
```
Comment (wdanilo) [inline lib/rust/parser/src/macros.rs:1]: We are not using plural names in modulee naes. This sghould be named macro (until its reservedword, is it?)

File: lib/rust/parser/src/macros.rs [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+    registry : Registry,
+    /// The logger for the macro resolver.
+    logger : Logger
+}
+
+impl<Logger:AnyLogger<Owned=Logger>> Resolver<Logger> {
+    /// Constructor.
+    pub fn new(macros:Vec<Definition>, parent_logger:&Logger) -> Self {
+        let registry      = Registry::from(macros);
+        let logger        = <Logger>::sub(parent_logger,"Resolver");
```
Comment (wdanilo) [inline lib/rust/parser/src/macros.rs]: spacing

File: lib/rust/parser/src/macros/definition.rs [hunk 17]
```diff
@@ -0,0 +1,87 @@
+//! Macro definitions in Enso.
+
+use crate::prelude::*;
+use crate::macros::literal::Literal;
+use itertools::Itertools;
+
+
+// ==================
```
Comment (wdanilo) [inline lib/rust/parser/src/macros/definition.rs:9]: nlns

File: lib/rust/parser/src/macros/definition.rs [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+/// [sections](`Section`). The sections are the most important portion of the macro definition, as
+/// they define the literal portions of the token stream on which the macro will match.
+#[derive(Clone,Debug,Default,Eq,PartialEq)]
+pub struct Definition {
+    /// The name of the macro definition.
+    ///
+    /// This is used both in error messages and for the purposes of nesting a macro within another
+    /// macro.
+    name : String,
+    /// The sections that make up the macro.
```
Comment (wdanilo) [inline lib/rust/parser/src/macros/definition.rs]: this doc is useless

File: lib/rust/parser/src/macros/definition.rs [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+// ==================
+
+/// A macro definition.
+///
+/// A macro definition consists of a name, which identifies the macro to users, and a list of
+/// [sections](`Section`). The sections are the most important portion of the macro definition, as
+/// they define the literal portions of the token stream on which the macro will match.
+#[derive(Clone,Debug,Default,Eq,PartialEq)]
+pub struct Definition {
+    /// The name of the macro definition.
```
Comment (wdanilo) [inline lib/rust/parser/src/macros/definition.rs]: this doc is useless

File: lib/rust/parser/src/macros/definition.rs [hunk 20]
```diff
@@ -0,0 +1,87 @@
...
+
+    /// Get the name of the macro.
+    pub fn name(&self) -> &str {
+        self.name.as_str()
+    }
+
+    /// Get the sections that make up this macro definition.
+    pub fn sections(&self) -> &[Section] {
+        self.sections.as_slice()
+    }
```
Comment (wdanilo) [inline lib/rust/parser/src/macros/definition.rs]: its ok, but why dont make the fields public ?

File: lib/rust/parser/src/macros/definition.rs [hunk 21]
```diff
@@ -0,0 +1,87 @@
...
+
+/// A section in a macro, representing both a literal section header to match against, and the
+/// tokens that the section contains.
+///
+/// The literal is the _most_ important portion of a section, as they are constants that allow the
+/// macro resolver to divide up the input token stream based on these constants.
+#[derive(Clone,Debug,Eq,PartialEq)]
+pub struct Section {
+    /// The literal that begins the section.
+    symbol : Literal,
```
Comment (wdanilo) [inline lib/rust/parser/src/macros/definition.rs]: the name is not telling about its semantic purpose. aybe it sohuld be named `begin_symbol` and then the docs can be removed too

File: lib/rust/parser/src/macros/definition.rs [hunk 22]
```diff
@@ -0,0 +1,87 @@
...
+    // TODO Pattern (later)
```
Comment (wdanilo) [inline lib/rust/parser/src/macros/definition.rs]: what does (later) mean here?

File: lib/rust/parser/src/macros/definition.rs [hunk 23]
```diff
@@ -0,0 +1,87 @@
...
+
+impl Section {
+    /// Constructor.
+    pub fn new(symbol:Literal) -> Self {
+        Self{ symbol }
+    }
+
+    /// Get a reference to the literal that heads the section.
+    pub fn literal(&self) -> &Literal {
+        &self.symbol
```
Comment (wdanilo) [inline lib/rust/parser/src/macros/definition.rs]: this name is not semanticaly self explanatory

File: lib/rust/parser/src/macros/definition.rs [hunk 24]
```diff
@@ -0,0 +1,87 @@
...
+    }
+}
+
+
+// === Macros ===
+
+// TODO Macro for section definition.
```
Comment (wdanilo) [inline lib/rust/parser/src/macros/definition.rs]: what is this todo? Neds better explanation

File: lib/rust/parser/src/macros/literal.rs [hunk 25]
```diff
@@ -0,0 +1,86 @@
...
+
+
+// ===============
+// === Literal ===
+// ===============
+
+/// The kinds of literal that can be the head of a macro section.
+#[derive(Clone,Debug,Eq,Hash,Ord,PartialEq,PartialOrd)]
+pub enum Literal {
+    Referent(String),
```
Comment (wdanilo) [inline lib/rust/parser/src/macros/literal.rs:18]: what is referent?

File: lib/rust/parser/src/macros/literal.rs [hunk 26]
```diff
@@ -0,0 +1,86 @@
...
+    Variable(String),
+    External(String),
```
Comment (wdanilo) [inline lib/rust/parser/src/macros/literal.rs:20]: what is extenral?

File: lib/rust/parser/src/macros/literal.rs [hunk 27]
```diff
@@ -0,0 +1,86 @@
...
+    Blank,
+    Operator(String),
+    Annotation(String)
```
Comment (wdanilo) [inline lib/rust/parser/src/macros/literal.rs:23]: what is annotation? I need example

File: lib/rust/parser/src/macros/registry.rs [hunk 28]
```diff
@@ -0,0 +1,136 @@
...
+
+    /// Get a reference to the root of the registry.
+    pub fn root(&self) -> &Tree {
+        &self.tree
+    }
+
+    /// Query the registry for a tree.
+    pub fn tree(&self, path:&[Literal]) -> Option<&Tree> {
+        self.tree.get(path)
+    }
```
Comment (wdanilo) [inline lib/rust/parser/src/macros/registry.rs]: this method has non explanatory name. ITs shoudl be "get", or "subtree". The same applies to method below

**Review by wdanilo — APPROVED** [review]

File: lib/rust/parser/src/data/tree.rs (hunk 10 above)
Comment (iamrecursion) [reply lib/rust/parser/src/data/tree.rs]: You can't use self as this is a type changing operation.

File: lib/rust/parser/src/data/tree.rs (hunk 11 above)
Comment (iamrecursion) [reply lib/rust/parser/src/data/tree.rs]: These aren't meant to be links as you can't link to parameters.

File: lib/rust/parser/src/macros.rs (hunk 15 above)
Comment (iamrecursion) [reply lib/rust/parser/src/macros.rs:1]: It is a reserved word unfortunately. That's why I used `macros`.

**Pull request merged**